benches/codecs/newline_bytes.rs

```diff
@@ -0,0 +1,90 @@
+use bytes::BytesMut;
+use criterion::{
+    criterion_group, measurement::WallTime, BatchSize, BenchmarkGroup, BenchmarkId, Criterion,
+    SamplingMode, Throughput,
+};
+use std::{fmt, time::Duration};
+use tokio_util::codec::Decoder;
+use vector::codecs::{self, BytesDeserializer, NewlineDelimitedDecoder};
+
+#[derive(Debug)]
+struct Param {
+    slug: &'static str,
+    input: BytesMut,
+    max_length: Option<usize>,
+}
+
+impl fmt::Display for Param {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.slug)
+    }
+}
+
+fn decoding(c: &mut Criterion) {
+    let mut group: BenchmarkGroup<WallTime> =
+        c.benchmark_group("vector::codecs::decoding::Decoder");
+    group.sampling_mode(SamplingMode::Auto);
+
+    for param in &[
+        Param {
+            slug: "no_max",
+            input: BytesMut::from(include_str!("moby_dick.txt")),
+            max_length: None,
+        },
+        Param {
+            slug: "small_max",
+            input: BytesMut::from(include_str!("moby_dick.txt")),
+            max_length: Some(10),
+        },
+    ] {
+        group.throughput(Throughput::Bytes(param.input.len() as u64));
+        group.bench_with_input(
+            BenchmarkId::new("newline_bytes", param),
+            &param,
+            |b, param| {
+                b.iter_batched(
+                    || {
+                        let framer = Box::new(
+                            param
+                                .max_length
+                                .map(|ml| NewlineDelimitedDecoder::new_with_max_length(ml))
+                                .unwrap_or(NewlineDelimitedDecoder::new()),
+                        );
+                        let deserializer = Box::new(BytesDeserializer::new());
+                        let decoder = codecs::Decoder::new(framer, deserializer);
+
+                        (Box::new(decoder), param.input.clone())
+                    },
+                    |(mut decoder, mut input)| loop {
+                        match decoder.decode_eof(&mut input) {
+                            Ok(Some(_)) => continue,
+                            Ok(None) => break,
+                            Err(_) => {
+                                unreachable!()
+                            }
+                        }
```

Comment:
Looking at https://github.com/vectordotdev/vector/blob/633a27c3c28bbc6cde3eeb34ff7c32651b42f12f/src/codecs/framing/character_delimited.rs#L133-L145 (`NewlineDelimitedDecoder` is just a `CharacterDelimitedCodec` with `\n` as the character).

It seems like the implementation (which `decode_eof` calls) returns `Ok(None)` the first time it fails to find a full line in `max_length` so I think this will `break` as soon as it hits a line over 10 characters.

I think the implementation may have a bug though. I don't think it should return `Ok(None)` unless it does actually hit the end of the buffer.

Consider the implementation of Tokio's `LinesCodec`: https://docs.rs/tokio-codec/latest/src/tokio_codec/lines_codec.rs.html#118-183.

@pablosichert may have more thoughts here.

Reply:
> Looking at
> 
> https://github.com/vectordotdev/vector/blob/633a27c3c28bbc6cde3eeb34ff7c32651b42f12f/src/codecs/framing/character_delimited.rs#L133-L145
> (`NewlineDelimitedDecoder` is just a `CharacterDelimitedCodec` with `\n` as the character).

Indeed, born out by the numbers as well.

Reply:
I opened https://github.com/vectordotdev/vector/pull/10568 to fix this behavior.

Reply:
Holy moly, good analysis on that one and embarrassing that it existed in the first place.

Reply:
I ran these benches with https://github.com/vectordotdev/vector/pull/10568 merged in. They looked like:


```
vector::codecs::decoding::Decoder/character_delimited_bytes/no_max
                        time:   [32.329 ms 32.401 ms 32.474 ms]
                        thrpt:  [36.207 MiB/s 36.289 MiB/s 36.370 MiB/s]
Found 2 outliers among 150 measurements (1.33%)
  1 (0.67%) low mild
  1 (0.67%) high mild
Benchmarking vector::codecs::decoding::Decoder/character_delimited_bytes/small_max: Warming up for 5.0000 s
Warning: Unable to complete 150 samples in 120.0s. You may wish to increase target time to 213.6s, enable flat sampling, or reduce sample count to 70.
vector::codecs::decoding::Decoder/character_delimited_bytes/small_max
                        time:   [18.303 ms 18.385 ms 18.473 ms]
                        thrpt:  [63.651 MiB/s 63.955 MiB/s 64.241 MiB/s]
Found 9 outliers among 150 measurements (6.00%)
  2 (1.33%) low severe
  2 (1.33%) low mild
  4 (2.67%) high mild
  1 (0.67%) high severe

Benchmarking vector::codecs::decoding::Decoder/newline_bytes/no_max: Warming up for 5.0000 s
Warning: Unable to complete 150 samples in 120.0s. You may wish to increase target time to 125.7s, enable flat sampling, or reduce sample count to 100.
vector::codecs::decoding::Decoder/newline_bytes/no_max
                        time:   [10.750 ms 10.781 ms 10.813 ms]
                        thrpt:  [108.74 MiB/s 109.06 MiB/s 109.38 MiB/s]
Found 8 outliers among 150 measurements (5.33%)
  1 (0.67%) low severe
  4 (2.67%) low mild
  3 (2.00%) high mild
vector::codecs::decoding::Decoder/newline_bytes/small_max
                        time:   [3.9817 ms 3.9917 ms 4.0022 ms]
                        thrpt:  [293.79 MiB/s 294.56 MiB/s 295.30 MiB/s]
Found 6 outliers among 150 measurements (4.00%)
  1 (0.67%) low severe
  1 (0.67%) low mild
  2 (1.33%) high mild
  2 (1.33%) high severe
```

My hunch is that the ones with the smallish max are still faster because they end up mostly discarding bytes though I would have thought the overhead of returning byte slices to be fairly small 🤔

Reply:
Okay, good. They're closer. Very, very good. Once this goes up and we have the delimiter adjusted to be a byte -- ref https://github.com/vectordotdev/vector/issues/10569 -- I'll experiment with https://github.com/vectordotdev/vector/pull/10550 again and we'll have at least micro-numbers available for seeing how that moves. I don't, as yet, actually have a hypothesis about what's up here, but we're limiting the scope of possibility. 

Fun problem.